synapse/events/snapshot.py

```diff
@@ -482,6 +477,38 @@ def _encode_state_dict(
     return [(etype, state_key, v) for (etype, state_key), v in state_dict.items()]
 
 
+def _encode_state_group_delta(
+    state_group_delta: Optional[
+        Dict[Tuple[Optional[int], Optional[int]], Optional[StateMap[str]]]
+    ]
+) -> Optional[
+    List[Tuple[Optional[int], Optional[int], Optional[List[Tuple[str, str, str]]]]]
+]:
+    if state_group_delta is None:
+        return None
+
+    state_group_delta_encoded = []
+    for key, value in state_group_delta.items():
+        state_group_delta_encoded.append((key[0], key[1], _encode_state_dict(value)))
+
+    return state_group_delta_encoded
+
+
+def _decode_state_group_delta(
+    input: Optional[
+        List[Tuple[Optional[int], Optional[int], Optional[List[Tuple[str, str, str]]]]]
+    ]
+) -> Optional[Dict[Tuple[Optional[int], Optional[int]], Optional[StateMap[str]]]]:
+    if input is None:
+        return None
+
+    state_group_deltas = {}
+    for element in input:
+        state_group_deltas[(element[0], element[1])] = _decode_state_dict(element[2])
+
+    return state_group_deltas
+
+
 def _decode_state_dict(
```

Comment:
Do we still use the `_{encode,decode}_state_dict` functions?

Reply:
Yes they are still used in a: the en/decoding of `state_delta_due_to_event` and b: as a helper method in `_{encode, decode}_state_group_deltas`